Pytorch kütüphanesi ile,kütüphanenin hazır aktivasyon fonksiyonlarını kullanmadan,

Tanh Fonksiyonu [ f(x)= (exp(x)-exp(-x) ) / (exp(x)+exp(-x) ) ]

Sigmoid Fonksiyonu [ f(x)= 1/( 1+exp(-x) ) ]

formüllerini kullanarak kendi yapay sinir ağımızı oluşturalım.

Yapılacaklar: 
* Forward propagation için ,input olarak x matrisi verilecek
* Matris tensöre çevrilecek

-Satırlar veriler(Sample)

-Kolonlar öznitelikler(Feature)

X=[
    
    1 2 3           
    4 5 6
          
          ]

* Hidden Layer=50 Nöron
* Kullanılacak Tanh ve Sigmoid fonksiyonu üst kısımdaki hücrede verilmiştir.

4.1-) 

Bu yapay sinir ağını çalıştırmadan önce pytorch için seed değerini 1 olarak ayarlıyoruz

In [1]:
import torch

# PyTorch'ta seed değeri 1 olarak ayarlandı
torch.manual_seed(1)

# Tanh ve sigmoid aktivasyon fonksiyonlarını kendimiz yazdık
def tanh(x):
    result=(torch.exp(x)-torch.exp(-x))/(torch.exp(x)+torch.exp(-x))
    return result

def sigmoid(x):
    result=1/(1+torch.exp(-x)) 
    return result

# Input X matrisi tensöre çevrildi
X = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float)

input_size=3
hidden_size=50
output_size=1

# Gizli katman için ağırlıklar ve bias değerleri tanımlandı
hidden_weights = torch.randn((input_size, hidden_size), dtype=torch.float)
hidden_bias = torch.randn((output_size, hidden_size), dtype=torch.float)

# Çıkış katmanı için ağırlıklar ve bias değerleri tanımlandı
output_weights = torch.randn((hidden_size, output_size), dtype=torch.float)
output_bias = torch.randn((1, 1), dtype=torch.float)

# İleri besleme işlemi gerçekleştirildi
hidden_layer = tanh(torch.mm(X, hidden_weights) + hidden_bias)
output_layer = sigmoid(torch.mm(hidden_layer, output_weights) + output_bias)

# Çıkış katmanındaki tek nöronun çıkışı yazdırıldı
print(output_layer)
 

tensor([[0.0498],
        [0.0075]])



Çıktı:

tensor([[0.0498],
        [0.0075]])

4.2-)


* Seed değerini öğrenci numaram olarak değiştirerek tekrar çalıştırıyorum

 -> torch.manual_seed(210401003)

In [20]:
import torch

# PyTorch'ta seed değeri ogrenci numaram olarak ayarlandı
torch.manual_seed(210401003)

# Tanh ve sigmoid aktivasyon fonksiyonlarını kendimiz yazdık
def tanh(x):
    result=(torch.exp(x)-torch.exp(-x))/(torch.exp(x)+torch.exp(-x))
    return result

def sigmoid(x):
    result=1/(1+torch.exp(-x)) 
    return result

# Input X matrisi tensöre çevrildi
X = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float)

input_size=3
hidden_size=50
output_size=1

# Gizli katman için ağırlıklar ve bias değerleri tanımlandı
hidden_weights = torch.randn((input_size, hidden_size), dtype=torch.float)
hidden_bias = torch.randn((output_size, hidden_size), dtype=torch.float)

# Çıkış katmanı için ağırlıklar ve bias değerleri tanımlandı
output_weights = torch.randn((hidden_size, output_size), dtype=torch.float)
output_bias = torch.randn((1, 1), dtype=torch.float)

# İleri besleme işlemi gerçekleştirildi
hidden_layer = tanh(torch.mm(X, hidden_weights) + hidden_bias)
output_layer = sigmoid(torch.mm(hidden_layer, output_weights) + output_bias)

# Çıkış katmanındaki tek nöronun çıkışı yazdırıldı
print(output_layer)

tensor([[0.9450],
        [0.8613]])


Çıktı:

tensor([[0.9450],
        [0.8613]])


## İki farklı seed değerinin çalışma zamanı olarak karşılaştırılması

In [21]:
import torch
import time

start_time = time.time()

# PyTorch'ta seed değeri 1 olarak ayarlandı
torch.manual_seed(1)

# Tanh ve sigmoid aktivasyon fonksiyonlarını kendimiz yazdık
def tanh(x):
    result=(torch.exp(x)-torch.exp(-x))/(torch.exp(x)+torch.exp(-x))
    return result

def sigmoid(x):
    result=1/(1+torch.exp(-x)) 
    return result

# Input X matrisi tensöre çevrildi
X = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float)

# Gizli katman için ağırlıklar ve bias değerleri tanımlandı
hidden_weights = torch.randn((3, 50), dtype=torch.float)
hidden_bias = torch.randn((1, 50), dtype=torch.float)

# Çıkış katmanı için ağırlıklar ve bias değerleri tanımlandı
output_weights = torch.randn((50, 1), dtype=torch.float)
output_bias = torch.randn((1, 1), dtype=torch.float)

# İleri besleme işlemi gerçekleştirildi
hidden_layer = tanh(torch.mm(X, hidden_weights) + hidden_bias)
output_layer = sigmoid(torch.mm(hidden_layer, output_weights) + output_bias)

# Çıkış katmanındaki tek nöronun çıkışı yazdırıldı
print(output_layer)

end_time = time.time()
print("Çalışma Süresi:",(end_time - start_time) )

tensor([[0.0498],
        [0.0075]])
Çalışma Süresi: 0.004000663757324219


In [22]:
import torch
import time

start_time = time.time()

# PyTorch'ta seed değeri 1 olarak ayarlandı
torch.manual_seed(210401003)

# Tanh ve sigmoid aktivasyon fonksiyonlarını kendimiz yazdık
def tanh(x):
    result=(torch.exp(x)-torch.exp(-x))/(torch.exp(x)+torch.exp(-x))
    return result

def sigmoid(x):
    result=1/(1+torch.exp(-x)) 
    return result

# Input X matrisi tensöre çevrildi
X = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float)

# Gizli katman için ağırlıklar ve bias değerleri tanımlandı
hidden_weights = torch.randn((3, 50), dtype=torch.float)
hidden_bias = torch.randn((1, 50), dtype=torch.float)

# Çıkış katmanı için ağırlıklar ve bias değerleri tanımlandı
output_weights = torch.randn((50, 1), dtype=torch.float)
output_bias = torch.randn((1, 1), dtype=torch.float)

# İleri besleme işlemi gerçekleştirildi
hidden_layer = tanh(torch.mm(X, hidden_weights) + hidden_bias)
output_layer = sigmoid(torch.mm(hidden_layer, output_weights) + output_bias)

# Çıkış katmanındaki tek nöronun çıkışı yazdırıldı
print(output_layer)

end_time = time.time()
print("Çalışma Süresi:",(end_time - start_time) )

tensor([[0.9450],
        [0.8613]])
Çalışma Süresi: 0.0035142898559570312


----------------------------------------------------------------------------------------------------------

Seed=1 

Sonuç:

tensor([[0.0498],
        [0.0075]])
        
Çalışma Süresi: 0.0039250850677490234

----------------------------------------------------------------------------------------------------------

Seed =210401003

Sonuç:

tensor([[0.9450],
        [0.8613]])

Çalışma Süresi: 0.00299835205078125

----------------------------------------------------------------------------------------------------------